glean-core/ios/sdk_generator.sh

```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# Glean SDK metrics build script.
+#
+# More about Glean at https://mozilla.github.io/glean
+#
+# This script generates metrics and pings as defined in user-provided files
+# and generates Swift code to be included in the final build.
+# It uses the `glean_parser`.
+# See https://mozilla.github.io/glean_parser/ for details.
+#
+# To use it in a Swift project, follow these steps:
+# 1. Import the `sdk_generator.sh` script into your project.
+# 2. Add your `metrics.yaml` and (optionally) `pings.yaml` to your project.
+# 3. Add a new "Run Script" build step and set the command to `bash $PWD/sdk_generator.sh`
+# 4. Add your definition files (`metrics.yaml`, `pings.yaml`) as Input Files for the "Run Script" step.
+# 5. Run the build.
+# 6. Add the files in the `Generated` folder to your project.
+# 7. Add the same files from the `Generated` folder as Output Files of the newly created "Run SCript" step.
+# 8. Start using the generated metrics.
+
+set -e
+
+GLEAN_PARSER_VERSION=1.8.2
+
+# This script runs a given Python module as a "main" module, like
+# `python -m module`. However, it first checks that the installed
+# package is at the desired version, and if not, upgrades it using `pip`.
+RUN_PYTHON_SCRIPT=$(cat <<'EOF'
+import importlib
+import subprocess
+import sys
+
+module_name = sys.argv[1]
+expected_version = sys.argv[2]
+
+try:
+    module = importlib.import_module(module_name)
+except ImportError:
+    found_version = None
+else:
+    found_version = getattr(module, '__version__')
+
+if found_version != expected_version:
+    subprocess.check_call([
+        sys.executable,
+        '-m',
+        'pip',
+        'install',
+        '--upgrade',
+        f'{module_name}=={expected_version}'
+    ])
+
+subprocess.check_call([
+    sys.executable,
+    '-m',
+    module_name
+] + sys.argv[3:])
+EOF
+)
+
+# When the special argument "internal" is passed, don't add a namespace import and also allow all reserved items.
+NAMESPACE=Glean
+INTERNAL=""
+if [ "$1" = "internal" ]; then
+  NAMESPACE=""
+  INTERNAL="--allow-reserved"
+fi
+
+pipenv --venv || pipenv install --three
```

Comment:
pipenv seems like overkill here, because we aren't actually using it to install any dependencies.  

Additionally, it's going to create a new virtual environment for every directory that this script is run from, i.e. one for Glean itself, and one for all the libraries and apps we have in the future that use Glean.  It's only 17MB each (so not nearly as bad as the 350MB each that we have from using Miniconda in Android land -- that's what we get for not needing to support Windows here...)

I think it might be better / more standard to:

1) Create a virtualenv in a single central location:

```
python3 -m venv ~/.local/glean/venv
```

2) Run glean parser from there...

```
~/.local/glean/venv/bin/python3 -c "$RUN_PYTHON_SCRIPT" ...
```

Reply:
I guess alternatively if you *wanted* to use `pipenv` you could include a Pipfile that sets `glean_parser==1.8.2` and have it pick that up, then we don't need to do any custom updating nonsense at all...

Reply:
Well, we do use `pip` after all.
The reason I reached for pipenv are that for some reason xcode includes some half-broken Python thingy, which is picked up first if I don't use pipenv. I can't run any python on that.

Additionally, I really don't like a project-specific script installing a global across-projects-used virtualenv. That is just calling for issues once anyone has multiple Glean-using projects, that might depend on different versions of the parser.
Plus, we can't even rely on anything outside of the project directory to exist or be writable.

I'll try again using venv directly and calling pip without the detour through a hacky script/